libtcmu.c

```diff
@@ -383,18 +383,192 @@ static void tcmu_cdb_debug_info(const struct tcmulib_cmd *cmd)
+{
+	struct tcmu_aio_entry *aio_entry;
+	struct tcmu_aio_queue *aio_wq = &dev->work_queue;
+
+	aio_entry = malloc(sizeof(*aio_entry));
+	if (!aio_entry) {
+		return -ENOMEM;
+	}
+
+	list_node_init(&aio_entry->entry);
+
+	aio_entry->cmd = cmd;
+	aio_entry->call = call;
+
+	/* cleanup push/pop not _really_ required here atm */
+	pthread_cleanup_push(_cleanup_mutex_lock, &aio_wq->aio_lock);
+	pthread_mutex_lock(&aio_wq->aio_lock);
+
+	list_add_tail(&aio_wq->aio_queue, &aio_entry->entry);
+	pthread_cond_signal(&aio_wq->aio_cond); // TODO: conditional
+
+	pthread_mutex_unlock(&aio_wq->aio_lock);
+	pthread_cleanup_pop(0);
+
+	return TCMU_ASYNC_HANDLED;
+}
+
+static int setup_aio_work_queue(struct tcmu_device *dev)
+{
+	int ret;
+	struct tcmu_aio_queue *aio_wq = &dev->work_queue;
+
+	list_head_init(&aio_wq->aio_queue);
+
+	ret = pthread_mutex_init(&aio_wq->aio_lock, NULL);
+	if (ret < 0) {
+		goto out;
+	}
+	ret = pthread_cond_init(&aio_wq->aio_cond, NULL);
+	if (ret < 0) {
+		goto cleanup_lock;
+	}
+
+	// TODO: >1 worker threads (per device via config)
+	ret = pthread_create(&aio_wq->aio_wq_thread, NULL, aio_work_queue, dev);
+	if (ret < 0) {
+		goto cleanup_cond;
+	}
+
+	return 0;
+
+cleanup_cond:
+	pthread_cond_destroy(&aio_wq->aio_cond);
+cleanup_lock:
+	pthread_mutex_destroy(&aio_wq->aio_lock);
+out:
+	return ret;
+}
+
+static void cleanup_aio_work_queue(struct tcmu_device *dev,
+				   bool cancel)
+{
+	int ret;
+	struct tcmu_aio_entry *curr, *next;
+	struct tcmu_aio_queue *aio_wq = &dev->work_queue;
+
+	if (cancel) {
+		cancel_thread(aio_wq->aio_wq_thread);
+	}
+
+	list_for_each_safe(&aio_wq->aio_queue, curr, next, entry) {
+		list_del(&curr->entry);
+		free(curr);
+	}
+
+	ret = pthread_mutex_destroy(&aio_wq->aio_lock);
+	if (ret != 0) {
+		tcmu_err("failed to destroy aio workqueue lock\n");
+	}
+
+	ret = pthread_cond_destroy(&aio_wq->aio_cond);
+	if (ret != 0) {
+		tcmu_err("failed to destroy aio workqueue cond\n");
+	}
+}
+
+static int setup_aio_tracking(struct tcmu_device *dev)
+{
+	int ret;
+	struct tcmu_track_aio *aio_track = &dev->track_queue;
+
+	list_head_init(&aio_track->track_queue);
+
+	ret = pthread_spin_init(&aio_track->track_lock, 0);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static void cleanup_aio_tracking(struct tcmu_device *dev)
+{
+	int ret;
+	struct tcmu_track_aio *aio_track = &dev->track_queue;
+
+	ret = pthread_spin_destroy(&aio_track->track_lock);
+	if (ret < 0) {
+		tcmu_err("failes to destroy track lock\n");
+	}
+}
+
+static void async_call_command(struct tcmu_device *dev,
+			       struct tcmulib_cmd *cmd)
+{
 	struct tcmulib_handler *handler = tcmu_get_dev_handler(dev);
 	struct tcmur_handler *r_handler = handler->hm_private;
-	struct pollfd pfd;
+
 	int ret;
+	int (*func)(struct tcmu_device *, struct tcmulib_cmd *) =
+		(r_handler->handle_cmd) ? : generic_handle_cmd;
+
+	if (r_handler->aio_supported) {
+		ret = func(dev, cmd);
+	} else {
+		ret = aio_schedule(dev, cmd, func);
```

Comment:
Since ```func``` is just derived from ```dev```, how about creating a helper function for the logic on lines 549-550 and eliminate the need to pass a function pointer to the ```aio_schedule``` method (and the need to track the function pointer in addition to tracking the device).